sdks/java/core/src/main/java/org/apache/beam/sdk/coders/MapCoder.java

```diff
@@ -149,11 +162,15 @@ public void registerByteSizeObserver(
       Map<K, V> map, ElementByteSizeObserver observer, Context context)
       throws Exception {
     observer.update(4L);
-    for (Entry<K, V> entry : map.entrySet()) {
-      keyCoder.registerByteSizeObserver(
-          entry.getKey(), observer, context.nested());
-      valueCoder.registerByteSizeObserver(
-          entry.getValue(), observer, context.nested());
+    Iterator<Entry<K, V>> entries = map.entrySet().iterator();
+    while (entries.hasNext()) {
+      Entry<K, V> entry = entries.next();
+      keyCoder.registerByteSizeObserver(entry.getKey(), observer, context.nested());
+      if (entries.hasNext()) {
+        valueCoder.registerByteSizeObserver(entry.getValue(), observer, context.nested());
+      } else {
+        valueCoder.registerByteSizeObserver(entry.getValue(), observer, context);
+      }
```

Comment:
Relatively complex changes. But as single-element maps are probably likely to be used in global combining, suspect it should be worth it.